src/flp/types.rs

```diff
@@ -419,55 +417,90 @@ impl<F: FftFriendlyFieldElement> Type for Histogram<F> {
     }
 }
 
-/// A sequence of counters. This type uses a neat trick from [[BBCG+19], Corollary 4.9] to reduce
-/// the proof size to roughly the square root of the input size.
+/// A sequence of integers in range `[0, 2^bits)`. This type uses a neat trick from [[BBCG+19],
+/// Corollary 4.9] to reduce the proof size to roughly the square root of the input size.
 ///
 /// [BBCG+19]: https://eprint.iacr.org/2019/188
 #[derive(Debug, PartialEq, Eq)]
-pub struct CountVec<F, S> {
+pub struct SumVec<F: FftFriendlyFieldElement, S> {
     range_checker: Vec<F>,
     len: usize,
+    bits: usize,
+    flattened_len: usize,
+    max: F::Integer,
     chunk_len: usize,
     gadget_calls: usize,
     phantom: PhantomData<S>,
 }
 
-impl<F: FftFriendlyFieldElement, S: ParallelSumGadget<F, BlindPolyEval<F>>> CountVec<F, S> {
-    /// Returns a new [`CountVec`] with the given length.
-    pub fn new(len: usize) -> Self {
+impl<F: FftFriendlyFieldElement, S: ParallelSumGadget<F, BlindPolyEval<F>>> SumVec<F, S> {
+    /// Returns a new [`SumVec`] with the bit width and vector length.
+    ///
+    /// # Errors
+    ///
+    /// * The length of the encoded measurement, i.e., `bits * len`, overflows addressible memory
+    /// * The bit width cannot be encoded, i.e., `bits` is larger than the number of bits in
+    /// required to encode field elements.
+    ///
+    /// Behavior is undefined if the size of `F::Integer` exceeds `F::ENCODED_SIZE`.
```

Comment:
I think this is backwards, and we only get into trouble if `F::Integer` has a smaller range than `F`. (assuming this is about calculation of `max`)

Reply:
Yes, this is about computing `max`. My claim is that the computation of `max` is guaranteed to not overflow if the size of `F::Integer` is less than or equal to `F::ENCODED_SIZE`, hence the behavior being undefined if this is not he case.

However, it occurred to me that we can easily get around this undefined behavior since `F::Integer` is `Sized`:

```rust
let limit = std::mem::size_of::<F::Integer>() * 8 - 1;
```